From a4f33b0df1b0255c53e01aae9c00ab654fc273ef Mon Sep 17 00:00:00 2001
From: Sandor Yu <R01008@freescale.com>
Date: Fri, 29 May 2015 17:07:04 +0800
Subject: [PATCH] MLK-10256-02 v4l2: Add v4l2_async_notifier_unregister
 recursively call

If anyone calls v4l2_async_notifier_unregister() recursively from
device_release_driver(), code will deadlock at list_lock, so unlock
list_lock when device_release_driver() called.

Signed-off-by: Sandor Yu <R01008@freescale.com>
---
 drivers/media/v4l2-core/v4l2-async.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/media/v4l2-core/v4l2-async.c b/drivers/media/v4l2-core/v4l2-async.c
index 85a6a34..a0dc460 100644
--- a/drivers/media/v4l2-core/v4l2-async.c
+++ b/drivers/media/v4l2-core/v4l2-async.c
@@ -213,7 +213,14 @@ void v4l2_async_notifier_unregister(struct v4l2_async_notifier *notifier)
 		v4l2_async_cleanup(sd);
 
 		/* If we handled USB devices, we'd have to lock the parent too */
+		/*
+		 * If anyone calls v4l2_async_notifier_unregister() recursively from
+		 * device_release_driver(), code will deadlock at list_lock,
+		 * so unlock list_lock when device_release_driver() called.
+		 */
+		mutex_unlock(&list_lock);
 		device_release_driver(d);
+		mutex_lock(&list_lock);
 
 		if (notifier->unbind)
 			notifier->unbind(notifier, sd, sd->asd);
-- 
1.8.0

